mesonbuild/dependencies/dev.py

```diff
@@ -505,6 +506,59 @@ def get_methods():
         return [DependencyMethods.SYSTEM]
 
 
+class JDKSystemDependency(ExternalDependency):
+    def __init__(self, environment: 'Environment', kwargs: T.Dict[str, T.Any]):
+        super().__init__('jdk', environment, kwargs)
+
+        m = self.env.machines[self.for_machine]
+
+        java_home = os.environ.get('JAVA_HOME')
+        if java_home is None:
+            mlog.debug('$JAVA_HOME is unset')
+            self.is_found = False
+            return
+
+        self.version = environment.detect_java_compiler(self.for_machine).version
+
+        platform_include_dir = JDKSystemDependency.__machine_info_to_platform_include_dir(m)
+        if platform_include_dir is None:
+            mlog.debug('Unknown platform include directory')
+            self.is_found = False
+            return
+
+        java_home_include = pathlib.Path(java_home) / 'include'
+
+        if self.clib_compiler.get_argument_syntax() == 'gcc':
+            self.compile_args.append(f'-I{java_home_include}')
+            self.compile_args.append(f'-I{java_home_include / platform_include_dir}')
+            self.is_found = True
+        elif self.clib_compiler.get_argument_syntax() == 'msvc':
+            self.compile_args.append(f'/I{java_home_include}')
+            self.compile_args.append(f'/I{java_home_include / platform_include_dir}')
+            self.is_found = True
+
+    @staticmethod
+    def get_methods():
+        return [DependencyMethods.SYSTEM]
+
+    @staticmethod
+    def __machine_info_to_platform_include_dir(m: MachineInfo) -> T.Optional[str]:
```

Comment:
It doesn't look like you've imported `MachineInfo`, I'd suggest adding
```python
from ..envconfig import MachineInfo
```
inside the `if T.TYPE_CHECKING:` block, then change this to a forward declaration (python < 3.7 does that with a string) so:
````python
def __function(m: 'MachineInfo') ...
```

Reply:
This is why the tests are failing, btw.